lib/MonadicRewrite.thy

```diff
@@ -403,72 +522,88 @@ lemma monadic_rewrite_gets_the_walk:
 
-lemma monadic_rewrite_gets_known:
-  "monadic_rewrite F E (\<lambda>s. f s = rv) (gets f >>= m) (m rv)"
-  by (simp add: monadic_rewrite_def exec_gets)
+lemma monadic_rewrite_refine_validE_R:
+  "\<lbrakk> monadic_rewrite F E P f g;
+     \<lbrace>P'\<rbrace> f \<lbrace>Q\<rbrace>, -;
+     F \<longrightarrow> no_fail P'' f \<rbrakk>
+   \<Longrightarrow> \<lbrace>P and P' and P''\<rbrace> g \<lbrace>Q\<rbrace>, -"
+  by (simp add: validE_R_def validE_def monadic_rewrite_refine_valid)
 
-lemma monadic_rewrite_name_pre:
-  "\<lbrakk> \<And>s. P s \<Longrightarrow> monadic_rewrite F E ((=) s) f g \<rbrakk>
-    \<Longrightarrow> monadic_rewrite F E P f g"
-  by (auto simp add: monadic_rewrite_def)
+text \<open>Specific failure flags and flag manipulation\<close>
 
-lemma monadic_rewrite_named_bindE:
-  "\<lbrakk> monadic_rewrite F E ((=) s) f f';
-    \<And>rv s'. (Inr rv, s') \<in> fst (f' s)
-      \<Longrightarrow> monadic_rewrite F E ((=) s') (g rv) (g' rv) \<rbrakk>
-    \<Longrightarrow> monadic_rewrite F E ((=) s) (f >>=E (\<lambda>rv. g rv)) (f' >>=E g')"
-  apply (rule monadic_rewrite_imp)
-   apply (erule_tac R="(=) s" and Q="\<lambda>rv s'. (Inr rv, s') \<in> fst (f' s)" in monadic_rewrite_bindE)
-    apply (rule monadic_rewrite_name_pre)
-    apply clarsimp
-   apply (clarsimp simp add: validE_R_def validE_def valid_def
-                      split: sum.split)
-  apply simp
-  done
+lemma monadic_rewrite_to_eq:
+  "monadic_rewrite False True \<top> f g \<Longrightarrow> f = g"
+  by (simp add: monadic_rewrite_def fun_eq_iff)
 
-lemmas monadic_rewrite_named_if
-    = monadic_rewrite_if[where Q="(=) s" and R="(=) s", simplified] for s
+lemma monadic_rewrite_transverse:
+  "\<lbrakk> monadic_rewrite False True Q c b; monadic_rewrite F E P a b \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (P and Q) a c"
+  by (auto simp add: monadic_rewrite_def)
 
-lemma monadic_rewrite_if_lhs:
-  "\<lbrakk> P \<Longrightarrow> monadic_rewrite F E Q b a; \<not> P \<Longrightarrow> monadic_rewrite F E R c a \<rbrakk>
-      \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (P \<longrightarrow> Q s) \<and> (\<not> P \<longrightarrow> R s))
-             (If P b c) a"
-  by (cases P, simp_all)
+lemma monadic_rewrite_weaken_failure:
+  "\<lbrakk> monadic_rewrite True True P f f'; no_fail P' f; no_fail Q' f' \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (P and P' and Q') f f'"
+  by (clarsimp simp: monadic_rewrite_def no_fail_def)
 
-lemma monadic_rewrite_to_eq:
-  "monadic_rewrite False True \<top> f g ==> f = g"
-  by (simp add: monadic_rewrite_def fun_eq_iff)
+lemma monadic_rewrite_is_valid:
+  "\<lbrakk> monadic_rewrite False False P' f f'; \<lbrace>P\<rbrace> do x <- f; g x od \<lbrace>Q\<rbrace> \<rbrakk>
+   \<Longrightarrow> \<lbrace>P and P'\<rbrace> do x <- f'; g x od \<lbrace>Q\<rbrace>"
+  by (fastforce simp: monadic_rewrite_def valid_def bind_def)
+
+text \<open>corres / refinement\<close>
 
 lemma corres_underlyingI:
```

Comment:
Would this rule and the following be better in `Corres_UL.thy` or something? Or is that bad because of imports?

Reply:
Very good point. I'll have a look.

Reply:
Done